### PR TITLE
adds bf16 nvfusion (INCOMPLETE, only to show diffs)

### PR DESCRIPTION
**Pull request by puririshi98**

### CODE REVIEW

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -72,6 +72,17 @@ Tensor autocast_to_fp16(const Tensor& self) {
   }
 }
 
+// If input tensor is fp32, cast it to fp16, otherwise leave it alone.
+// (this is intended to be used internally by the JIT autocast implementation)
+Tensor autocast_to_bf16(const Tensor& self) {
```
Comment (csarofeen) [inline aten/src/ATen/native/TensorConversions.cpp:77]: Do we support auto casting in AMP for bf16?

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp [hunk 2]
```diff
@@ -716,7 +765,7 @@ void SegmentedFusion::finalize() {
       TensorView* cast_tv = nullptr;
       // Insert cast ops for this tv if we haven't done so.
       if (cast_tv_it == fp32_to_fp16_cast_map.end()) {
-        cast_tv = castIntermediateValueInCompleteFusion(
+        cast_tv = castIntermediateValueInCompleteFusionFP16(
             complete_fusion_.get(), edge_tv);
         fp32_to_fp16_cast_map[edge->val->as<TensorView>()] = cast_tv;
       } else {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:804]: This doesn't look right.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp [hunk 3]
```diff
@@ -726,6 +775,25 @@ void SegmentedFusion::finalize() {
       // Update the edge to use the fp16 version
       edge->val = cast_tv;
 
+      // Mark the groups for update later
+      affected_group_set.insert(edge->from);
+      affected_group_set.insert(edge->to);
+    }
+    if (force_bf16_tv_set_.count(edge_tv)) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: If in both bf16 set and fp16 set throw an error.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 4]
```diff
@@ -1345,6 +1357,22 @@ class IrParser {
           },
           nullptr,
           nullptr);
+
+    }
+
+    {
+      auto ptr_op = getOperatorForLiteral(
+          "aten::autocast_to_bf16(Tensor(a) self) -> Tensor(a)");
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1368]: Does pytorch support autocast with bf16?

File: torch/csrc/jit/codegen/cuda/runtime/fp16_support.cu [hunk 5]
```diff
@@ -17,6 +17,7 @@ struct __align__(2) __half {
   unsigned short __x;
 };
 
+
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/runtime/fp16_support.cu]: nit remove extra spaces.

**Review by csarofeen — COMMENTED** [review]
Overall seems okay. We need tests of course for bf16, so you should look in `test/cpp/jit/test_gpu.cpp` and `test/test_jit_cuda_fuser.py` for places to also enable bf16 testing.

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (puririshi98) [reply aten/src/ATen/native/TensorConversions.cpp:77]: in upstream master we do. I made a PR to get it into upstream a while back along with some follow up fixes 
https://github.com/pytorch/pytorch/pull/61002
https://github.com/pytorch/pytorch/pull/63798
https://github.com/pytorch/pytorch/pull/64122

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 4 above)
Comment (puririshi98) [reply torch/csrc/jit/codegen/cuda/parser.cpp:1368]: > in upstream master we do. I made a PR to get it into upstream a while back along with some follow up fixes
> [pytorch#61002](https://github.com/pytorch/pytorch/pull/61002)
> [pytorch#63798](https://github.com/pytorch/pytorch/pull/63798)
> [pytorch#64122](https://github.com/pytorch/pytorch/pull/64122)

**Pull request closed without merging**